### ISSUE TITLE
Ethernet connection issue with Raspberry pi 3 Model B

### ISSUE THREAD

**Issue by agangal93** [OP-TEE/optee_os#3335]
I am trying to connect the Raspberry pi with Optee OS to the ethernet without using NFS boot. The pi is booting and I am able to run xtest successfully. I am using a stable release branch 3.4.0.

```
23304 subtests of which 0 failed
93 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
# dmesg | grep smsc95
[    0.000000] Kernel command line: console=tty0 console=ttyS0,115200 root=/dev/mmcblk0p2 rw rootfs=ext4 smsc95xx.macaddr=b8:27:eb:fc:00:a0 ignore_loglevel dma.dmachans=0x7f35 rootwait 8250.nr_uarts=1 elevator=deadline fsck.repair=yes bcm2708_fb.fbwidth=1920 bcm2708_fb.fbheight=1080 vc_mem.mem_base=0x3ec00000 vc_mem.mem_size=0x40000000 dwc_otg.fiq_enable=0 dwc_otg.fiq_fsm_enable=0 dwc_otg.nak_holdoff=0
[    0.376943] usbcore: registered new interface driver smsc95xx
[    2.985722] smsc95xx v1.0.6
[    3.040699] smsc95xx 1-1.1:1.0 eth0: register 'smsc95xx' at usb-3f980000.usb-1.1, smsc95xx USB 2.0 Ethernet, b8:27:eb:fc:00:a0
# ifconfig eth0 up
[  729.454909] smsc95xx 1-1.1:1.0 eth0: hardware isn't capable of remote wakeup
# [  731.398374] smsc95xx 1-1.1:1.0 eth0: link up, 100Mbps, full-duplex, lpa 0xC1E1
ifconfig -a
eth0      Link encap:Ethernet  HWaddr B8:27:EB:FC:00:A0  
          inet6 addr: fe80::ba27:ebff:fefc:a0/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:17 errors:0 dropped:0 overruns:0 frame:0
          TX packets:7 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:1408 (1.3 KiB)  TX bytes:642 (642.0 B)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:667 errors:0 dropped:0 overruns:0 frame:0
          TX packets:667 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:133236 (130.1 KiB)  TX bytes:133236 (130.1 KiB)
```

I followed the steps mentioned in https://github.com/OP-TEE/optee_os/issues/1398 to get the ip configured using dhpc service, but I am stuck at this step
/etc/init.d/udhcpc start
since I cannot find the udhcpc service in init.d

```
# ls -al /etc/init.d/

total 32
drwxr-xr-x    2 root     root          4096 Oct 16  2019 .
drwxr-xr-x    7 root     root          4096 Jan  1 00:00 ..
-rwxr-xr-x    1 root     root           656 Oct 14  2019 S01logging
-rwxr-xr-x    1 root     root          1272 Oct 14  2019 S20urandom
-rwxr-xr-x    1 root     root          1082 Oct 16  2019 S30optee
-rwxr-xr-x    1 root     root           438 Oct 14  2019 S40network
-rwxr-xr-x    1 root     root           423 Oct 14  2019 rcK
-rwxr-xr-x    1 root     root           408 Oct 14  2019 rcS
```

Contents of /etc/network/interfaces file
```
# interface file auto-generated by buildroot

auto lo
iface lo inet loopback
auto eth0
iface eth0 inet dhcp
metric 100
```
1. Has this been updated in the latest release or I am missing some important step here?
2. Once connected, are the standard socket programming apis available to transfer data from the pi to the host PC/server? Can this be achieved from a TA, or we need to transfer the data back to the client side before transmitting over the network.

Thank you for your time.

Regards,
Akshay Gangal

**Comment by agangal93** [comment]
I tried the ethernet connection again with the latest release 3.7.0. I am still seeing the same issue. Do we need to change something during the build to include the udhcpc service?
@jbech-linaro , @igoropaniuk  it would be really helpful if you can provide some inputs on this issue. I have followed the steps mentioned in #1398 by @igoropaniuk 

Regards,
Akshay Gangal

**Comment by igoropaniuk** [comment]
@agangal93 udhcpc.sh just wraps invocation of udhcpc busybox tool.
```
#!/bin/sh
case "$1" in
  start)
	printf "Starting network (udhcpc): "
	udhcpc >> /var/log/udhcpc.log 2>&1
	[ $? = 0 ] && echo "OK" || echo "FAIL"
	;;
  stop)
	printf "Stopping network (udhcpc): "
	killall udhcpc
	[ $? = 0 ] && echo "OK" || echo "FAIL"
	;;
  restart|reload)
	"$0" stop
	"$0" start
	;;
  *)
	echo "Usage: $0 {start|stop|restart}"
	exit 1
esac

exit $?

```
Have you tried running just `udhcpc` instead? Does it work for you?

**Comment by agangal93** [comment]
Hi @igoropaniuk,

Thanks a lot. I tried just ```udhcpc``` but it keeps sending discover without response. I found a similar issue in https://github.com/OP-TEE/optee_os/issues/2903.

![Network_1](https://user-images.githubusercontent.com/43253458/68043382-89062800-fcab-11e9-9f18-f95ba83f8722.png)

I configured the IP address manually as mentioned in https://github.com/OP-TEE/optee_os/issues/2903 after which I was able to fix this issue. Networking works fine now(I am able to ping it from my host system), with the only issue being that I have to set it up every time after boot. 
1. Is it possible to add a new script which can automate this?

![Network_2](https://user-images.githubusercontent.com/43253458/68043645-219ca800-fcac-11e9-8d35-e29727015881.png)

2. Once connected, are the standard socket programming apis available to transfer data from the pi to the host PC/server? Can this be achieved from a TA, or we need to transfer the data back to the client side before transmitting over the network

Thanks a lot for your time.
